### PR TITLE
#26 deploy and display default template

### LINKED ISSUE #26: Deploy and display defaultTemplate contract
Acceptance criteria:
* the DefaultTemplate contract is deployed as part of the deploy.ts script and ownership is transferred to the Manager contract
* the template is shown in the UI

### PR DESCRIPTION
**Pull request by ricCap**
closes #26 
* fix broken CI pipeline
* deploy the template, assign it to the manager and display it in the UI
* update rollup plugin to fix local `vite` issues
* build types for both ethers (hardhat tests) and web3 (dapp)
  * dapp types are under `src/types`
  * hardhat types are still under `typechain-types`
* stop tracking compiled contracts/types and add them as a postinstall task (after `npm install`)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,14 +15,16 @@ jobs:
       - name: Setup node
         uses: actions/setup-node@v3
         with:
-          node-version: "16"
+          node-version: '16'
 
       - name: Install Dependencies
+        env:
+          MNEMONIC: 'fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake fake'
```
Comment (ricCap) [inline .github/workflows/ci.yml:22]: we should update the hardhat config to have a default if this is not set. I'll create an issue

File: contracts/projects/DefaultProjectTemplate.sol [hunk 2]
```diff
@@ -11,12 +11,13 @@ import "@openzeppelin/contracts/utils/Counters.sol";
 pragma solidity ^0.8.9;
 
 /** @dev Default project template */
-contract DefaultProject is IProjectTemplate, ERC721, ERC721Enumerable, Ownable {
+abstract contract DefaultProjectTemplate is IProjectTemplate, ERC721, ERC721Enumerable, Ownable {
```
Comment (ricCap) [inline contracts/projects/DefaultProjectTemplate.sol:14]: made this abstract

File: contracts/projects/ExampleProjectTemplate.sol [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+
+import "./DefaultProjectTemplate.sol";
+
+pragma solidity ^0.8.9;
+
+/** @dev Example project template */
+contract ExampleProjectTemplate is DefaultProjectTemplate {
```
Comment (ricCap) [inline contracts/projects/ExampleProjectTemplate.sol:8]: Example project template

File: scripts/deploy.ts [hunk 4]
```diff
@@ -1,14 +1,30 @@
...
+  await managerContract.addProjectTemplate(exampleProjectTemplateContract.address)
+  console.log(`Example template deployed to ${exampleProjectTemplateContract.address} and added to manager`)
+
+  // Store addresses into a file the app can use
+  const addresses: DeployedAddresses = {
+    Manager: managerContract.address,
+    ExampleProjectTemplate: exampleProjectTemplateContract.address,
+  }
+
+  fs.writeFileSync('./src/addresses.json', JSON.stringify(addresses, null, 2), 'utf8')
```
Comment (ricCap) [inline scripts/deploy.ts:27]: now we store the addresses in a file (imported by the app), no need to manual copy paste of the addresses anymore

File: src/addresses.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+  "Manager": "0x98c045Ff750E39aa404eB5d34432ef0581880003",
```
Comment (ricCap) [inline src/addresses.json:2]: this is generated by the deploy script

File: src/app/Navbar.tsx [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
 
 import { getI18N } from './i18n'
 import * as constants from './constants'
 import managerABI from '../../artifacts/contracts/Manager.sol/Manager.json'
+import { Manager } from '../types/contracts/Manager'
+import { ExampleProjectTemplate } from '../types/contracts/projects'
 
 export let kit: contractkit.ContractKit
-export let managerContract: Contract
+export let managerContract: Manager
```
Comment (ricCap) [inline src/app/Navbar.tsx:16]: this should be improved as I'd like to have a single object that contains all contracts instances, will be done in the future

File: src/app/Navbar.tsx [hunk 7]
```diff
@@ -159,7 +161,14 @@ async function _connect(props: INavbarProps, web3: Web3) {
   const accounts = await kit.web3.eth.getAccounts()
   kit.defaultAccount = accounts[0]
 
-  managerContract = new kit.web3.eth.Contract(managerABI.abi as unknown as AbiItem, constants.ManagerContractAddress)
+  managerContract = new kit.web3.eth.Contract(
+    managerABI.abi as AbiItem[],
+    constants.addresses.Manager,
+  ) as unknown as Manager
```
Comment (ricCap) [inline src/app/Navbar.tsx:167]: now we can use the types in `src/types`, so it's easier to call the contracts functions

File: src/app/constants.ts [hunk 8]
```diff
@@ -1,4 +1,6 @@
+import { DeployedAddresses } from '../../scripts/deploy'
```
Comment (ricCap) [inline src/app/constants.ts:1]: I need to move this type, probably it is not even necessary

File: package.json [hunk 9]
```diff
@@ -1,18 +1,25 @@
...
     "build:vite": "vite build",
     "build:css": "npx tailwindcss -i ./src/index.css -o ./dist/output.css",
     "build:contract": "npx hardhat compile",
+    "build:typechains": "npm run build:typechains-web3 && npm run build:typechains-ethers",
+    "build:typechains-web3": "npx hardhat typechain",
+    "build:typechains-ethers": "npx typechain --target=\"ethers-v5\" --out-dir typechain-types \"artifacts/contracts/**/*[!dbg].json\"",
     "lint": "npm run lint:sol",
     "lint:sol": "solhint --config ./.solhint.json --max-warnings 0 \"contracts/**/*.sol\"",
+    "clean": "npm run clean:typechains-ethers && npm run clean:typechains-web3 && rm -rf ./artifacts ./cache ./coverage ./coverage.json",
+    "clean:typechains-ethers": "rm -rf typechains-ethers",
```
Comment (ricCap) [inline package.json:14]: types

**Review by MattiaMolon — APPROVED** [review]

**Pull request merged**